shards/sched.go

```diff
@@ -27,30 +30,73 @@ import (
 // the indexer proceeds. To do this we have an exclusive process which will be
 // the only one running. This is like a Lock on a RWMutex, while a normal
 // search is a RLock.
+//
+// ### Co-operatively limit long running searches
+//
+// Some searches are slow. Either due to a hard to execute search query (can't
+// use trigram index) or a large number of results. We want to support this
+// use case while still allowing interactive queries to be fast.
+//
+// ### No tuneables
+//
+// We want to avoid the need to tune the scheduler depending on the workload /
+// instance. As such we use a simple design whose inputs are time and number
+// of CPUs.
+//
+// ## Design
+//
+// We use semaphores to limit the number of running processes. An exclusive
+// process acquires the full semaphore. A process represents something which
+// has acquired on the semaphore. Every process is either fast or slow. A
+// process starts as fast, but is downgraded to slow after a period of
+// time. Downgrading relies on a process co-operatively deciding to downgrade.
+//
+// We intentionally keep the algorithm simple, but have a general interface to
+// allow improvements as we learn more.
 type scheduler struct {
 	throttle *semaphore.Weighted
+
+	// capacity is the max concurrent searches we allow.
 	capacity int64
+
+	// interactiveDuration is how long we run a search query at interactive
+	// priority before downgrading it to a batch/slow query.
+	interactiveDuration time.Duration
 }
 
 func newScheduler(capacity int64) *scheduler {
 	return &scheduler{
 		throttle: semaphore.NewWeighted(capacity),
 		capacity: capacity,
+
+		interactiveDuration: 5 * time.Second,
```

Comment:
As discussed yesterday will make this a hidden tuneable while we experiment with it. Also will serve as an escape hatch for large customers in case things go wrong.